### DISCUSSION TITLE
Python code minification tool

### DISCUSSION THREAD

**Discussion by createdbyjurand** [micropython/micropython#15451] [Ideas]
It would be great to have some tool to minify python code before sending it to the board.
It would save some space and we could have lots of comments in out source files during development.
Removing comments and white spaces would really be enough for the start.

**Comment by robert-hh** [comment]
What stops our from using your preferred minification tool and integrate it in your workflow for uploading files?

**Comment by peterhinch** [comment]
The simplest approach is to cross-compile: see [mpy-cross](https://github.com/micropython/micropython/tree/master/mpy-cross).

**Comment by sosi-deadeye** [comment]
Minify is used in JS world. Pythoneers don't do this. You can compile Python code and you can compile Micropython code.

The use of MPY is the right approach.

**Comment by createdbyjurand** [comment]
Oh wow, nice, thanks for the answers :) 
Yeah, I am from JS world.
I saw couple of minification tools out there for python:

- https://github.com/dflook/python-minifier
- https://github.com/liftoff/pyminifier

so I thought it is a thing.
Both are outdated and do not work with latest Python. Some are using libs that were deprecated. And that's the reason for this post.
I guess I was doing it wrong, because I was pushing `py` files on to the board.
Didn't know that MP can work on precompiled bytecode.
`mpy-cross` is changing everything.
Thanks :)

**Reply by glenn20** [reply]
I also like to liberally use type hints in my micropython code knowing that it is not bloating my production code because I always deploy it precompiled :)

**Comment by createdbyjurand** [comment]
I thought `mpy-cross` can be installed like `esptool` or `ampy`, I work on Windows and don't have `make`. 
So, would it be possible to have prebuild version in the repo?
https://github.com/micropython/micropython/tree/master/mpy-cross
Or make it installable like `esptool`?
Or am I missing something?

Or is it the same as:
- https://adafruit-circuit-python.s3.amazonaws.com/index.html?prefix=bin/mpy-cross/
- https://github.com/zestyping/mpy-cross?tab=readme-ov-file

**Reply by andrewleech** [reply]
You can `pip install mpy-cross` on windows (and most other platforms) to get a ready to run binary copy of mpy-cross

**Reply by createdbyjurand** [reply]
Ooooh it works!!! Oh man, you made my day :) Thank you! I fell so happy that I want to compile everything now hahaha

**Comment by ctimmer** [comment]
Here's another one :
https://pyminifier3.readthedocs.io/en/latest/

I tried a minifier a while back and it worked great. Cut the source by 2/3. Probably would have done better if I used comments. 😀

Curt

**Comment by shariltumin** [comment]
I use pyminify from https://pypi.org/project/python-minifier/

You can install it with
```
$ pip install python-minifier
```

The documentation can be found at https://dflook.github.io/python-minifier/

It is not inconceivable that you might want to obfuscate your code and reduce code size by reusing variable names. There are many reasons why you might want to do this.

I use pyminify and then mpy-cross on finished scripts.

**Reply by createdbyjurand** [reply]
But that does not work on python 3.12 :(
'python-minifier currently supports Python 2.7 and Python 3.3 to 3.11'

**Reply by Josverl** [reply]
True, python 3.12 support is being worked on, there were a few commits this week to address that.

Also, IN CI/CD that is a minor thing to work around, using pipx, actions, or hatch environments.

Ref: https://github.com/dflook/python-minifier/issues/98

**Comment by vshymanskyy** [comment]
[VIperIDE](https://viper-ide.org) is able to minify (and prettify) the source file.
- minification is done by running `python-minifier` (locally in-browser)
- prettification is done by `Ruff` formatter

Please note that these operations are destructive (however you can undo them in the editor view)

![image](https://github.com/user-attachments/assets/73931db6-8332-4348-8039-ada83233dc02)
